index-evolution/src/test/java/fax/play/smoke/MinorsForNoDowntimeUpgradesTest.java

```diff
@@ -0,0 +1,205 @@
+     * The difference is, that it seems for non-analyzed fields it could be possible to do the migration in one step
+     * on the other hand, I am not sure whether this works correctly as currently I don't know how to check whether
+     * the query used the index or not
+     */
+    @Test
+    void testAddNonAnalyzedIndexOnExistingField() {
+        // VERSION 1
+        RemoteCache<String, Model> cache = cacheProvider
+                .init(new CacheDefinition(CACHE1_NAME, "Model3"), Schema3A.INSTANCE)
+                .getCache(CACHE1_NAME);
+
+        // Create VERSION 1 entities
+        ModelUtils.createModel1Entities(cache, 5, ModelUtils.createModelA(1));
+
+        // Check there is only one with 3 in name field (Query that doesn't use index)
+        doQuery("FROM Model3 WHERE name LIKE '%3%'", cache, 1);
+
+        cache = cacheProvider.updateSchemaAndGet(Schema3B.INSTANCE)
+                .getCache(CACHE1_NAME);
+
+        // TODO: No index schema update nor reindex needed, is it correct?
+        //  Can I somehow check whether Infinispan used the index or not?
+
+        // Create second version entities
+        ModelUtils.createModel1Entities(cache, 5, ModelUtils.createModelB(2));
+
+        // check there are two entities with 3 in name field
+        doQuery("FROM Model3 WHERE name LIKE '%3%'", cache, 2);
+    }
+
+    /**
+     * Effectively the same as {@link BasicsForNoDowntimeUpgradesTest#testAddAIndexOnExistingFieldThatWasNotUsedInAnyQueryBefore}
+     * but in this case we want to add index on a field that was previously used in a query
+     */
+    @Test
+    void testAnalyzedIndexAddedInMoreSteps() {
+        // VERSION 1
+        RemoteCache<String, Model> cache = cacheProvider
+                .init(new CacheDefinition(CACHE1_NAME, "Model3"), Schema3A.INSTANCE)
+                .getCache(CACHE1_NAME);
+
+        // Create VERSION 1 entities
+        ModelUtils.createModel1Entities(cache, 5, ModelUtils.createModelA(1));
+        doQuery("FROM Model3 WHERE name LIKE '%3%'", cache, 1);
+
+        // VERSION 2
+        // Note: this version needs to be backward compatible with the old version
+        // Update to second schema
+        cache = cacheProvider.updateSchemaAndGet(Schema3D.INSTANCE)
+                .getCache(CACHE1_NAME);
+
+        // Index schema needs to be updated
+        cacheProvider.updateIndexSchema(CACHE1_NAME);
+
+        // Create second version entities
+        ModelUtils.createModel1Entities(cache, 5, ModelUtils.createModelD(2));
+
+        // Make query that count with both older version and newer version
+        // TODO: This doesn't work: ISPN005003: Exception reported java.lang.IllegalStateException: Unexpected condition type (FullTextTermExpr): PROP(nameIndexed):'*3*'
+        //   this is already reported https://issues.redhat.com/browse/ISPN-8584
+        doQuery("FROM Model3 WHERE (entityVersion = 1 AND name LIKE '%3%') OR (entityVersion >= 2 AND nameIndexed : '*3*')", cache, 2);
```

Comment:
Hmm, I disagree a little bit here. Of course, it would be possible to do the filtering on the client side, however, I see three problems with this:

1. Pagination basically doesn't work if we want to paginate a query that has non-indexed field in it
2. Loading entities from storage into Keycloak just throw them away is something we want to avoid as much as possible to save network utilization
3. At the moment, to workaround this issue we created an index for each field in all entities that result in a lot of indexes.  We don't have thorough performance results yet, but I would say creating an index is not a cheap operation, therefore we should be careful about creating the indexes and use them only in situations where it really makes sense. As an example, we have a boolean user field `enabled`. The majority of users have set it to true, there are only a few that are disabled. Now, the only option at the moment is to create an index for this even though it consists of `true for almost each object. 

Another and probably more serious problem is pagination. It would not work

Reply:
Micheal I open a topic here: https://infinispan.zulipchat.com/#narrow/stream/118645-infinispan/topic/ISPN-8584.20hybrid.20perdicate.